### PR TITLE
ROX-26252: Simplify operator install scripts

### PR DESCRIPTION
**Pull request by porridge**
### Description

Needed to enable installation of konflux-built operator index: https://issues.redhat.com/browse/ROX-26252.

I thought adding a third flavor of the same template file was too much, so I folded them into one. This in turn required changing the scripts to accept `v`-prefixed version strings rather than blindly prepending the `v` at the last moment, which breaks when the index image tag does not follow stackrox tag conventions. I renamed some variables to make more sense, and introduced new make variables that allow overriding things as needed (see the bottom of this description for an example).

As for recommended way or reviewing: individual commits explain each step of the refactoring, although reading it all in one go is also a sensible approach, since a few later changes modify code already changed by earlier ones.

### User-facing documentation

- [x] CHANGELOG is updated **OR** update is not needed
- [x] [documentation PR](https://spaces.redhat.com/display/StackRox/Submitting+a+User+Documentation+Pull+Request) is created and is linked above **OR** is not needed
- [ ] Announce to the team that the arguments expected by `operator/hack/olm-operator*sh` have changed, as have the override variables for some `operator/Makefile` targets, in the unlikely case that someone calls these by hand.

### Testing and quality

- [x] the change is production ready: the change is GA or otherwise the functionality is gated by a feature flag
- [x] CI results are inspected

#### Automated testing

- [ ] added unit tests
- [ ] added e2e tests
- [ ] added regression tests
- [ ] added compatibility tests
- [x] modified existing tests

#### How I validated my change

- [x] upstream, midstream case: CI - see comments below
- [x] upstream: from workstation: with `make deploy-via-olm` and with `make deploy-previous-via-olm && make test-upgrade`
- [x] downstream:
```
make -C operator deploy-via-olm \
  ROX_PRODUCT_BRANDING=RHACS_BRANDING \
  INSTALL_CHANNEL=stable \
  INSTALL_VERSION=v4.5.2 \
  INDEX_IMG_TAG=ocp-v4-12-9504fd9990aaba298f90478b7345e28bfc1a06d7-fast-amd64
```

### CODE REVIEW

**Comment by porridge** [comment]
The powervs/ibmcloudz tests all timed out, but at least on the latest revision the operator deployment seems to have gone fine.

**Comment by porridge** [comment]
> The powervs/ibmcloudz tests all timed out, but at least on the latest revision the operator deployment seems to have gone fine.

I took a closer look at the `ibmcloudz` failures, and they all seem to experience multiple QA test failures, each of which triggers artifact collection that is rather slow (for example just collecting secrets from `kube-system` takes ~6 minutes) - each such iteration takes 20-30 minutes, so this quickly adds up to blow through the 6h deadline.

However:
1. I don't see a single successful run of any of these jobs for as long as prow history goes (April)
2. In general this does not seem to be related to changes in this PR

**Comment by porridge** [comment]
Regarding the `powervs` jobs, half of them failed to create a cluster, and the other half failed for reasons unrelated to this change, mosly due to missing images, e.g.:
```
    	  Container status: [ContainerState(running=null, terminated=null, waiting=ContainerStateWaiting(message=Container image "quay.io/rhacs-eng/qa-multi-arch:busybox-1-33-1" is not present with pull policy of Never, reason=ErrImageNeverPull, additionalProperties={}), additionalProperties={})]
    	  Container status: [ContainerState(running=null, terminated=null, waiting=ContainerStateWaiting(message=Container image "quay.io/rhacs-eng/qa-multi-arch:nginx-1.12@sha256:72daaf46f11cc753c4eab981cbf869919bd1fee3d2170a2adeac12400f494728" is not present with pull policy of Never, reason=ErrImageNeverPull, additionalProperties={}), additionalProperties={})]
    	  Container status: [ContainerState(running=null, terminated=null, waiting=ContainerStateWaiting(message=Container image "quay.io/rhacs-eng/qa-multi-arch:nginx-1-19-alpine" is not present with pull policy of Never, reason=ErrImageNeverPull, additionalProperties={}), additionalProperties={})]
    	  Container status: [ContainerState(running=null, terminated=null, waiting=ContainerStateWaiting(message=Container image "quay.io/rhacs-eng/qa-multi-arch:trigger-policy-violations-most-v1" is not present with pull policy of Never, reason=ErrImageNeverPull, additionalProperties={}), additionalProperties={})]
```

It's likely that this message is relevant: `orchestratormanager.OrchestratorManagerException: The deployment did not start or reach replica ready state - if this job uses image prefetch check that this image is in the jobs prefetch list e.g. qa-tests-backend/scripts/images-to-prefetch.txt - quay.io/rhacs-eng/qa-multi-arch:trigger-policy-violations-most-v1`

For:
- [pull-ci-stackrox-stackrox-master-ocp-stable-scanner-v4-ebpf-qa-e2e-tests](https://prow.ci.openshift.org/job-history/gs/test-platform-results/pr-logs/directory/pull-ci-stackrox-stackrox-master-ocp-stable-scanner-v4-ebpf-qa-e2e-tests)
- [pull-ci-stackrox-stackrox-master-ocp-stable-scanner-v4-nongroovy-e2e-tests](https://prow.ci.openshift.org/job-history/gs/test-platform-results/pr-logs/directory/pull-ci-stackrox-stackrox-master-ocp-stable-scanner-v4-nongroovy-e2e-tests)
- [pull-ci-stackrox-stackrox-master-ocp-stable-scanner-v4-qa-e2e-tests](https://prow.ci.openshift.org/job-history/gs/test-platform-results/pr-logs/directory/pull-ci-stackrox-stackrox-master-ocp-stable-scanner-v4-qa-e2e-tests)
- [pull-ci-stackrox-stackrox-master-ocp-stable-scanner-v4-ui-e2e-tests](https://prow.ci.openshift.org/job-history/gs/test-platform-results/pr-logs/directory/pull-ci-stackrox-stackrox-master-ocp-stable-scanner-v4-ui-e2e-tests)
- the failures seem unrelated, and I'm not able to find a single passing case in the past

For [pull-ci-stackrox-stackrox-master-osd-aws-qa-e2e-tests](https://prow.ci.openshift.org/job-history/gs/test-platform-results/pr-logs/directory/pull-ci-stackrox-stackrox-master-osd-aws-qa-e2e-tests) the failures look like  typical unrelated flakes. [pull-ci-stackrox-stackrox-master-rosa-qa-e2e-tests](https://prow.ci.openshift.org/job-history/gs/test-platform-results/pr-logs/directory/pull-ci-stackrox-stackrox-master-rosa-qa-e2e-tests) seems similar. The last successful run of the first of these jobs was back in August.

File: operator/hack/olm-operator-install.sh
Comment (msugakov) [inline operator/hack/olm-operator-install.sh:39]: Let's make the example complete:
```suggestion
    echo -e "Example:\n\t$0 -d index-test quay.io/rhacs-eng/stackrox-operator-index v3.70.1 v3.70.0 latest" >&2
```

File: operator/hack/common.sh
Comment (msugakov) [inline operator/hack/common.sh:38]: It seems that `operator_channel` is redefined from line 27.

File: operator/hack/common.sh
Comment (msugakov) [inline operator/hack/common.sh:40]: Please move `install_plan_approval` and `starting_csv` before or after this part that declares and assigns `disable_security_context_config` and `has_scc_key` variables. Both `install_plan_approval` and `starting_csv` just look out of place here.

File: operator/hack/common.sh
Comment (msugakov) [inline operator/hack/common.sh:48]: So strictly speaking, users of midstream images won't get `csv_version` they desire even if they provide something on the command line, right?

I think it's better to fail with error in case `install_plan_approval == Automatic && csv_version != index_image_tag` because users won't get what they asked for.

File: operator/hack/common.sh
Comment (msugakov) [inline operator/hack/common.sh:56]: May I ask you to break variables one per line? This way, it will be easier to review changes in this command in the future.
```suggestion
    INDEX_IMAGE_TAG="${index_image_tag}" \
    STARTING_CSV="${starting_csv}" \
    NAMESPACE="${operator_ns}" \
    OPERATOR_CHANNEL="${operator_channel}" \
    INDEX_IMAGE_REPO="${index_image_repo}" \
    DISABLE_SECURITY_CONTEXT_CONFIG="${disable_security_context_config}" \
    INSTALL_PLAN_APPROVAL="${install_plan_approval}" \
```

**Review by msugakov — COMMENTED** [review]
Only partial review today. Mostly cosmetics.

File: operator/hack/common.sh
Comment (porridge) [reply operator/hack/common.sh:48]: > So strictly speaking, users of midstream images won't get csv_version they desire even if they provide something on the command line, right?

Right, but that is also the case now. For example on current `master` [this setting](https://github.com/stackrox/stackrox/blob/db45ed40fd0f4d82e32c2c647380b93bd0637a6a/tests/e2e/lib.sh#L189) and following `export` has no impact. The `OPERATOR_VERSION` value does get plumbed by the `Makefile` through to `olm-operator-install.sh` and the `envsubst` call, but the old midstream template does not even use this variable. So the `operator.version` field in `iib.json` is effectively unused - the latest version in the `stable` channel gets installed.

> I think it's better to fail with error

Perhaps, but this would:
- be a change in semantics,
- require syncing with the team responsible for Power/Z (w.r.t. what is really intended)
- arguably be outside the scope of this PR
- slow down the konflux effort

So I'd prefer not to change this here.

File: operator/hack/common.sh
Comment (porridge) [reply operator/hack/common.sh:40]: Interestingly my IDE does not highlight the `local` keyword unless it's at the top of the function (preceded only by other declarations and value assignments). But perhaps this is only a parser's deficiency since shellcheck seems happy 🤞🏻 .

File: operator/hack/common.sh
Comment (msugakov) [reply operator/hack/common.sh:47]: Since the change already touches P/Z, I think, we need to do some cleanups to make things less confusing. I'm sure you spent some time figuring things out and I also spent time trying to understand how things work as part of this review. There's redundant stuff that must go away to not take away time from us and colleagues in the future.

Let me try suggest something that does not change semantics that much.

File: operator/hack/olm-operator-install.sh
Comment (msugakov) [inline operator/hack/olm-operator-install.sh:49]: I know you did not touch that, but `USE_MIDSTREAM_IMAGES` is redundant concept in the entire `operator/` directory. If we are to make things better, let's get rid of it and introduce more direct argument `INSTALL_PLAN_APPROVAL=(Manual|Automatic)`.

Anticipating the suggestion to do it in a later PR, my vote would be: no, let's try make it here (or in a stacked PR on top).

My further suggestion would be, in `tests/e2e/lib.sh`, to call `operator/hack/olm-operator-install.sh` directly instead of `make ... deploy-via-olm` in case of midstream images code branch.

https://github.com/stackrox/stackrox/blob/ed67f09fb494682790d623a5dc6863cf2cce141e/tests/e2e/lib.sh#L185-L194

The reason is that we override everything anyway. Why bother pushing things from the Makefile in a confusing way if we can call the thing directly. There will be a need to define `KUTTL` env variable but that's not critical given that we know the binary goes to a fixed path `operator/tools/kuttl`.

File: tests/e2e/lib.sh
Comment (msugakov) [inline tests/e2e/lib.sh:194]: Following my suggestions, this redundant `INSTALL_VERSION` can be simply removed as well as its corresponding `operator.version` https://github.com/stackrox/stackrox/blob/ed67f09fb494682790d623a5dc6863cf2cce141e/operator/midstream/iib.json#L10-L13

By the way, `operator.channel` can also be deleted from that file and hardcoded here, up to you.

File: operator/Makefile
Comment (msugakov) [inline operator/Makefile:16]: My suggestions should allow not introducing `INSTALL_VERSION` and `INSTALL_CHANNEL` variables here which in my opinion make things more confusing.

Also, I think `CSV_VERSION` can be deleted and `INDEX_IMG_TAG` be kept as the only thing. Perhaps as

```
# Tag of operator bundle and index images is version prefixed with "v".
INDEX_BUNDLE_IMG_TAG ?= v$(VERSION)
```

File: operator/Makefile
Comment (msugakov) [inline operator/Makefile:383]:
```suggestion
	$(SILENT)replaced_version=$$(sed -E -n 's/^[[:space:]]*replaces:[[:space:]]*[^.]+\.(v.*)$$/\1/p' build/bundle/manifests/rhacs-operator.clusterserviceversion.yaml) ;\
```

**Review by msugakov — COMMENTED** [review]
`INSTALL_VERSION` and `INSTALL_CHANNEL` introduced in the Makefile worry me. I don't like them as they decrease clarity and scripts become more convoluted. Here's my suggestion how deal with them.

File: operator/hack/olm-operator-install.sh
Comment (porridge) [reply operator/hack/olm-operator-install.sh:49]: Good idea about `$INSTALL_PLAN_APPROVAL`. Done.

However I'm quite strongly against calling anything in `operator/hack` from outside `operator/`. The makefile targets serve as the interface and give us a layer of indirection. For example I'd really prefer to avoid chasing uses of `KUTTL=` around if an when we migrate to chainsaw.

File: tests/e2e/lib.sh
Comment (porridge) [reply tests/e2e/lib.sh:194]: Ha, I dug a bit deeper and it turns out my claims about uselessness of this version designation for midstream-using jobs were not accurate. Version is required and used by the `nurse_deployment_until_available` function. So it needs to be specified and needs to match the most recent version in the channel of the iib image, otherwise installation will fail. As such it's best to keep this data together in `ibb.yaml`.

File: operator/Makefile
Comment (porridge) [reply operator/Makefile:16]: I'd like to keep `CSV_VERSION` as a simple way to specify both the tag and the version to install in one go.

File: operator/hack/common.sh
Comment (msugakov) [inline operator/hack/common.sh:5]: One more step is to pass `INSTALL_PLAN_APPROVAL` to the only function that needs it - to `apply_operator_manifests()`. Fewer globals simplify reading and understanding.

File: operator/hack/common.sh
Comment (msugakov) [inline operator/hack/common.sh:90]: Sorry, I did not mean that and you took my suggestion a bit too far. I meant to rename `csv_version` to `starting_csv_version` in `apply_operator_manifests` (and its call chain with `operator/hack/olm-operator-install.sh`) but not in other functions of `operator/hack/common.sh`.

It makes sense to keep calling it `csv_version` in `check_version_tag`, `approve_install_plan` and so on.

**Review by msugakov — COMMENTED** [review]
Quick notes from looking at interdiff. I'm not done reviewing yet.

File: tests/e2e/lib.sh
Comment (msugakov) [reply tests/e2e/lib.sh:195]: Good catch, but if that's the case and `operator.version` must match the latest version in the IIB index, we don't need Auto approval conditional at all. We can always deploy with Manual approval. That would make code simpler, wouldn't it?

File: operator/Makefile
Comment (msugakov) [inline operator/Makefile:10]:
```suggestion
# The operator version prefixed with "v" for ClusterServiceVersion resource and bundle and index tooling.
```

File: operator/Makefile
Comment (msugakov) [inline operator/Makefile:13]: Please move `INDEX_IMG_TAG` closer to `INDEX_IMG_BASE`. Also, I suggest revising description a bit
```
# INDEX_IMG_BASE and INDEX_IMG_TAG variables can be overridden when calling deploy*-via-olm targets, e.g. to deploy catalogs from IIB or other externally-built indexes.
```

File: operator/Makefile
Comment (msugakov) [inline operator/Makefile:16]: If we are to keep `INSTALL_VERSION` and `INSTALL_CHANNEL`, I suggest making their designation more clear.
At least something along the lines of
```suggestion
# Overridable operator version and channel to install for (deploy|upgrade)*-via-olm targets.
INSTALL_VERSION ?= $(CSV_VERSION)
INSTALL_CHANNEL ?= latest
```

If you can come up with more specific names, it will be better.

File: operator/hack/olm-operator-install.sh
Comment (msugakov) [reply operator/hack/olm-operator-install.sh]: Shell files calling Makefile calling shell files. I don't think this is a great design to stick to but if the kuttl thing is important for you, I would not insist.

File: tests/e2e/lib.sh
Comment (porridge) [reply tests/e2e/lib.sh:195]: Yes, it also seems so to me. @mdafsanhossain can you comment here on why the Auto approval option was introduced in the first place? Would it hurt if we standardized on Manual?

File: operator/hack/common.sh
Comment (porridge) [reply operator/hack/common.sh:5]: The `main` function in `olm-operator-install.sh` also uses it.

File: operator/hack/common.sh
Comment (porridge) [reply operator/hack/common.sh:90]: It seems my IDE does not know about `local` and renames all occurrences in the file. Let's keep it as is.

Comment (msugakov) [reply operator/hack/common.sh]: I added this comment because I believe renaming in other functions makes things more confusing. I did not expect that keeping it as-is would be a good option.

File: operator/hack/common.sh
Comment (msugakov) [reply operator/hack/common.sh:5]: Yes, that's my point: it has to be lifted as explicit argument to `olm-operator-install.sh` as well.

File: tests/e2e/lib.sh
Comment (pratham-m) [reply tests/e2e/lib.sh:195]: Switching to `manual` approval should also work. We might have to add code to approve the subscription changes (if its not already handled).
As regards to the parameters in `stackrox/operator/midstream/iib.json`, we wanted to have the parameters configurable so that a PR can be created to test a specific build.

File: operator/hack/common.sh
Comment (porridge) [reply operator/hack/common.sh:5]: > it has to be lifted as explicit argument

I agree _it would be nice_ to eradicate env variables. But I don't think a "has to be lifted" is in the scope of this PR given it's not something I'm introducing here. Our CI-related scripts are full of env variable dependencies that I absolutely despise.  And I'm all for applying the boy-scout rule, but I think this is getting a bit too far.

File: operator/hack/common.sh
Comment (porridge) [reply operator/hack/common.sh]: OK, done.

File: tests/e2e/lib.sh
Comment (porridge) [reply tests/e2e/lib.sh:195]: @pratham-m what exactly do you mean by "approve the subscription changes"? In the current flow the subscription does not get changed, I think?
There is an operator upgrade workflow, that is being exercised by the operator-e2e test jobs, but I don't think such job is configured for power/Z...

Comment (porridge) [reply tests/e2e/lib.sh:194]: Discussed offline. I got rid of the automatic mode.

File: operator/hack/common.sh
Comment (porridge) [reply operator/hack/common.sh]: The variable is gone now.

File: operator/hack/common.sh
Comment (msugakov) [inline operator/hack/common.sh:47]: Minor and optics: let's please put `INDEX_IMAGE_REPO` before `INDEX_IMAGE_TAG`. How about this reordering?
```suggestion
    NAMESPACE="${operator_ns}" \
    INDEX_IMAGE_REPO="${index_image_repo}" \
    INDEX_IMAGE_TAG="${index_image_tag}" \
    STARTING_CSV="rhacs-operator.${starting_csv_version}" \
    OPERATOR_CHANNEL="${operator_channel}" \
```

This way values follow the order of `operator/hack/olm-operator-install.sh` arguments. Not strictly necessary but makes it nicer.

File: operator/hack/olm-operator-install.sh
Comment (msugakov) [inline operator/hack/olm-operator-install.sh:25]: [nit, just noticed] If I read this correctly, "${6:-}" makes an argument optional but the following usage and the script proceeds as if the arguments are required.

I think this should be:
```suggestion
    local -r operator_ns="${1}"
    local -r index_image_repo="${2}"
    local -r index_image_tag="${3}"
    local -r starting_csv_version="${4}"
    local -r operator_channel="${5}"
```

File: tests/e2e/run-scanner-v4.bats
Comment (msugakov) [inline tests/e2e/run-scanner-v4.bats:401]: I just got to check this file hence this comment comes so late.

1. The use of `deploy_stackrox_operator` here and the use of `make -C operator deploy-previous-via-olm` below suggests this test file was never meant to run with midstream images and so by using `deploy_stackrox_operator` here we deceive ourselves that this _may_ work for P/Z whereas this actually doesn't. Would the tests still feel fine if we do the following?
```suggestion
    CSV_VERSION="v${OPERATOR_VERSION_TAG}" make -C operator deploy-via-olm
```
I think, consistency would be better than relying on the thin wrapper from `tests/e2e/lib.sh`.
Sorry, I know you did not touch _this_ aspect.

2. What makes me add comment here in the first place is again less than obvious mechanics of `CSV_VERSION` override. Would you be ok to swap it for `INDEX_IMG_TAG` and `INSTALL_VERSION`?

Comment (porridge) [reply tests/e2e/run-scanner-v4.bats:401]: The problem is that deploy_stackrox_operator does a couple more things (depends on the `kuttl` target and sets a couple of other env variables) and I _really_ don't think this is the right time to go and shave _that_ yak.

**Comment by porridge** [comment]
I agree the CSV_VERSION name is not ideal. But rather than have users repeat the version when overriding, I changed to overriding VERSION instead of CSV_VERSION. I hope this is something you could live with 😉 
PTAL @msugakov

File: operator/Makefile
Comment (msugakov) [inline operator/Makefile:9]: Optional suggestion:
```suggestion
# The operator version prefixed with "v" for ClusterServiceVersion resource, tagging bundle and index images and related tooling.
```

File: tests/e2e/run-scanner-v4.bats
Comment (msugakov) [reply tests/e2e/run-scanner-v4.bats]: Hm. Isn't there a problem that
* `deploy-previous-via-olm`
* `upgrade-via-olm`
* `upgrade-dirty-tag-via-olm`

all depend on `kuttl` but
* `deploy-via-olm`
* `deploy-dirty-tag-via-olm`

don't depend on it?

I mean the mere fact that I need to ask this and you need to spend time answering this is a sign to me that _that_ yak should have not been let grow hair from the beginning. From glancing at git history, I'm almost sure it was not intentional to have some targets depend on `kuttl` and others omit it but now I anyway have to guess why and waste your time on questions like this.

----

The absence of `REGISTRY_PASSWORD` and `REGISTRY_USERNAME` variables set in `Upgrade multi-namespace installation` test which calls `make -C operator deploy-previous-via-olm` directly makes me think that either tests are order-dependent and this test would fail should it be executed independently (or first in the batch) or that these variables are actually not needed for these scanner V4 tests.
Another yak that we let to happen?

----

There's no need to address these things in this PR but I would suggest you consider a follow-up to add `kuttl` as a dependency of `deploy-via-olm` and `deploy-dirty-tag-via-olm`.

**Review by msugakov — APPROVED** [review]
I find that using `VERSION` instead of `CSV_VERSION` in those places is indeed less confusing.

File: operator/hack/olm-operator-install.sh
Comment (msugakov) [inline operator/hack/olm-operator-install.sh:37]: Looking at this `apply_operator_manifests` call again, I would suggest you to consider using bash named arguments. Optional for this PR.
See https://redhat-internal.slack.com/archives/CELUQKESC/p1692118560975079

**Comment by msugakov** [comment]
Oh. One thing is that prow CI is very unhappy somehow. I did not look there. I hope you did and it's not related to the change.

File: operator/hack/olm-operator-install.sh
Comment (porridge) [reply operator/hack/olm-operator-install.sh:37]: Given this function has just one caller, and all args are required, I'm not seeing the benefit, TBH.
Especially given that in the caller there would be a lot of repetition:
`apply_operator_manifests "operator_ns=${operator_ns}" "index_image_repo=${index_image_repo}" "index_image_tag=${index_image_tag}" ...`

**Comment by porridge** [comment]
I looked at the failures and they all seem unrelated.

**Pull request merged**
